cryptoprotocol/src/main/java/de/upb/crypto/incentive/cryptoprotocol/model/Token.java

```diff
@@ -1,4 +1,135 @@
+        for(int i = 0; i < h.length; i++)
+        {
+            this.commitmentBases[i] = h[i]; // note: due to 0-based indexing, the bases' indices are off by one wrt paper
+        }
+
+        // initializing respective object variables with other parameters
+        this.myRemainderClassRing = myRemClassRing;
+        this.signatureScheme = sigScheme;
+        this.verificationKey = vKey;
+        this.userSecretKey = USK;
+        this.encryptionSecretKey = esk;
+        this.doubleSpendRandomness0 = dsrnd0;
+        this.doubleSpendRandomness1 = dsrnd1;
+        this.points = myRemainderClassRing.getZeroElement(); // initially, token contains no points
+        this.z = z;
+        this.t = t;
+
+        // compute the token-representing Pedersen commitment for the first time
+        this.updateToken();
+    }
+
+    /**
+     * recomputes the Pedersen commitment representing the token from the group elements and exponents (and also updates its plaintext representation)
+     * (see chapter 4 of 2020 incentive systems paper)
+     */
+    public void updateToken()
+    {
+        // note that indices of bases are off by one wrt paper (zero-based indexing)
+        // TODO get rid of getInteger calls
+        this.token = this.commitmentBases[0].pow(this.userSecretKey.getInteger()).op(
+                this.commitmentBases[1].pow(this.encryptionSecretKey.getInteger()).op(
+                        this.commitmentBases[2].pow(this.doubleSpendRandomness0.getInteger()).op(
+                                this.commitmentBases[3].pow(this.doubleSpendRandomness1.getInteger()).op(
+                                        this.commitmentBases[4].pow(this.points.getInteger()).op(
+                                                this.commitmentBases[5].pow(this.z.getInteger()).op(
+                                                        this.commitmentBases[6].pow(this.t.getInteger())
+                                                )
+                                        )
+                                )
+                        )
+                )
+        );
+
+        tokenPlainText = new GroupElementPlainText(this.token);
+    }
+
+
+    /**
+     * sets the certificate for the token.
+     * @param cert new certificate
+     */
+    public void setCertificate(SPSEQSignature cert) throws SPSEQException
+    {
+        this.certificate = cert;
+
+        // immediately throw exception if signature not valid for token group element
+        if(!signatureScheme.verify(this.tokenPlainText, this.certificate, this.verificationKey)); // token plaintext needed for verification (API reasons)
+        {
+            throw new SPSEQException("token was associated with an invalid certificate");
+        }
+    }
+
+    /**
+     * computes "dumb token" from mathematical representation of token
+     * @return serialized version of token
+     */
+    public Representation serializeToken()
+    {
+        // TODO
```

Comment:
Do you want to implement this before merge? Might be a good idea :)

Reply:
Done! Thanks for the reminder